CondFormats/GeometryObjects/interface/PDetGeomDesc.h

```diff
@@ -2,41 +2,25 @@
 #define CondFormats_PDetGeomDesc_h
 
 #include "CondFormats/Serialization/interface/Serializable.h"
-#include "Geometry/VeryForwardGeometryBuilder/interface/DetGeomDesc.h"
-#include "DetectorDescription/DDCMS/interface/DDTranslation.h"
-#include "DetectorDescription/DDCMS/interface/DDRotationMatrix.h"
 
 #include <vector>
 #include <string>
 
 class PDetGeomDesc {
 public:
-  struct Item {
-    Item() = default;
-    Item(const DetGeomDesc* const geoInfo) {
```

Comment:
It looks like one way to break the cyclic dependency is to remove this constructor and replace it with a helper function doing the same operations but in a different package.

Reply:
eg, where the code was before (as the class member data are public..)

Reply:
I wonder why this change even went in

Reply:
A priori is nicer to have things done in the constructor once and for all, than having a constructor doing nothing, and setting data members latter.
Then if there is a circular dependency in the packages, that makes it indeed not worth it, but it is a pity this does not get detected before.
How can this be detected?

Reply:
In the precious code for example, as expected, we find later down the line... 
PDetGeomDesc* pdet = new PDetGeomDesc;

Reply:
Which means that all built-in types are still not value-initialized there.

Of course, if there had been instead, 
PDetGeomDesc* pdet = new PDetGeomDesc() ,
at least data members would have been zero-initialized, which would already have been better.

Reply:
> In the previous code for example, as expected, we find later down the line...
PDetGeomDesc* pdet = new PDetGeomDesc;
> 
> Which means that all built-in types are still not value-initialized there.
>
> Of course, if there had been instead,
> PDetGeomDesc* pdet = new PDetGeomDesc() ,
>  at least built-in data members would have been zero-initialized, which would already have been better.

I'm not following. Both `new PDetGeomDesc;` and `new PDetGeomDesc();` call the default constructor of `PDetGeomDesc`. Both old an new code initialize `PDetGeomDesc::container_` as an empty vector.

Reply:
Yes because that s a std::vector
That's not the case for the built-in types: float, int, etc

Reply:
Correct, but the way the default constructor of the class whose members are built-in types does not matter for the initialization of these members. What matters is how such members are initialized (or not) in the constructor (or in the member declaration).

(Also `PDetGeomDesc` does not have any built-in type members)

Reply:
Ha yes wait the float, int, etc, are data members of Item anyway, not of PDetGeomDesc.

Reply:
> but the way the default constructor of the class whose members are built-in types does not matter for the initialization of these members

I take it a bit back that for a class without a user provided default constructor, whether the constructor is called with or without parentheses actually matters, and does dictate whether the member built-in types are zero-initialized. (the actual initialization rules are a bit complex
https://en.cppreference.com/w/cpp/language/default_initialization
https://en.cppreference.com/w/cpp/language/value_initialization
)

Reply:
@markotel, yes exactly, basically if PDetGeomDesc had built int types, there would not have been zero-initialized.
In general I tend to do what can be done in the constructor, instead of delaying (in terms of object life timeline) and having this type of situation. 
But I understand CondFormats/GeometryObjects is a special package, so we cannot simply add a constructor there without having a mess. Not worth the effort ;p

Reply:
You can (and I’d argue should) have a constructor, just not one using a DetGeomDesc. Instead, pass the values need to set the actual member data. That guarantees everything is properly initialized and avoids unnecessary coupling of packages.

Reply:
@Dr15Jones Ha thanks, good idea, that's the cleanest approach.